File: javascript/partition.js
var is_partitioned = function(seq, func) {
  for (var i = 0; i < seq.length; i++) {
    if (!func(seq[i])) {
      break;
    }
  }
  for (; i < seq.length; i++)
    if (func(seq[i])) return false;
  return true;
};

var partition_point = function(seq, func) {
  for (var i = 0; i < seq.length; i++)
    if (!func(seq[i])) return i;
  return i;
};

var partition = function(seq, func) {
  pivot = partition_point(seq, func);
  var tmp;
  for (var i = pivot; i < seq.length; i++) {
    if (func(seq[i])) {
      tmp = seq[pivot];
      seq[pivot] = seq[i];
      seq[i] = tmp;
      pivot += 1;
    }
  }
};

var partition_copy = function(seq, func) {
  res = [];
  var len = seq.length;
  for (var i = 0; i < len; i++) {
    if (func(seq[i])) {
      res.unshift(seq[i]);
    }
    else {
      res.push(seq[i]);
    }
  }
  return res;
};

var stable_partition = function(seq, func) {
  var len = seq.length;
  var tmp;
  for (var i = 0; i < len; i++) {
    if (!func(seq[i])) {
      tmp = Array.prototype.splice.call(seq, i, 1);
      Array.prototype.push.call(seq, tmp[0]);
    }
  }
};

module.exports = {
  is_partitioned: is_partitioned,
  partition_point: partition_point,
  partition: partition,
  partition_copy: partition_copy,
  stable_partition: stable_partition
};

File: lua/partition.lua
local function partition_point(seq, func)
    for i = 1, #seq do
        if not func(seq[i]) then return i end
    end
    return #seq + 1
end

local function is_partitioned(seq, func)
    local pivot = partition_point(seq, func)
    for i = pivot, #seq do
        if func(seq[i]) then return false end
    end
    return true
end

local function partition(seq, func)
    local pivot = partition_point(seq, func)
    for i = pivot, #seq do
        if func(seq[i]) then
            seq[i], seq[pivot] = seq[pivot], seq[i]
            pivot = pivot + 1
        end
    end
end

local function partition_copy(seq, func)
    local copy = {}
    for i = 1, #seq do
        if func(seq[i]) then
            table.insert(copy, 1, seq[i])
        else
            table.insert(copy, seq[i])
        end
    end
    return copy
end

local function stable_partition(seq, func)
    local seq_size = #seq
    local idx = 1
    for time = 1, seq_size do
        if not func(seq[idx]) then
            table.insert(seq, seq[idx])
            table.remove(seq, idx)
        else
            idx = idx + 1
        end
    end
end

return {
    is_partitioned = is_partitioned,
    partition_point = partition_point,
    partition = partition,
    partition_copy = partition_copy,
    stable_partition = stable_partition
}
